pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java

```diff
@@ -119,52 +118,58 @@ public ConsumerBuilderImpl(PulsarClientImpl client, Schema<T> schema) {
         if (conf.isRetryEnable() && conf.getTopicNames().size() > 0) {
             TopicName topicFirst = TopicName.get(conf.getTopicNames().iterator().next());
-            String retryLetterTopic =
-                    topicFirst + "-" + conf.getSubscriptionName() + RetryMessageUtil.RETRY_GROUP_TOPIC_SUFFIX;
-            String deadLetterTopic =
-                    topicFirst + "-" + conf.getSubscriptionName() + RetryMessageUtil.DLQ_GROUP_TOPIC_SUFFIX;
+            AtomicReference<String> retryLetterTopic =
+                    new AtomicReference<>(topicFirst + "-" + conf.getSubscriptionName()
+                            + RetryMessageUtil.RETRY_GROUP_TOPIC_SUFFIX);
+            AtomicReference<String> deadLetterTopic =
+                    new AtomicReference<>(topicFirst + "-" + conf.getSubscriptionName()
+                            + RetryMessageUtil.DLQ_GROUP_TOPIC_SUFFIX);
 
             //Issue 9327: do compatibility check in case of the default retry and dead letter topic name changed
             String oldRetryLetterTopic = topicFirst.getNamespace() + "/" + conf.getSubscriptionName()
                     + RetryMessageUtil.RETRY_GROUP_TOPIC_SUFFIX;
             String oldDeadLetterTopic = topicFirst.getNamespace() + "/" + conf.getSubscriptionName()
                     + RetryMessageUtil.DLQ_GROUP_TOPIC_SUFFIX;
-            try {
-                if (client.getPartitionedTopicMetadata(oldRetryLetterTopic)
-                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
-                    retryLetterTopic = oldRetryLetterTopic;
-                }
-                if (client.getPartitionedTopicMetadata(oldDeadLetterTopic)
-                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
-                    deadLetterTopic = oldDeadLetterTopic;
-                }
-            } catch (InterruptedException | TimeoutException e) {
-                return FutureUtil.failedFuture(e);
-            } catch (ExecutionException e) {
-                return FutureUtil.failedFuture(e.getCause());
-            }
-
-            if (conf.getDeadLetterPolicy() == null) {
-                conf.setDeadLetterPolicy(DeadLetterPolicy.builder()
-                                        .maxRedeliverCount(RetryMessageUtil.MAX_RECONSUMETIMES)
-                                        .retryLetterTopic(retryLetterTopic)
-                                        .deadLetterTopic(deadLetterTopic)
-                                        .build());
+            applyDLQConfig = client.getPartitionedTopicMetadata(oldRetryLetterTopic)
+                    .thenAccept(metadata -> {
+                        if (metadata.partitions > 0) {
+                            retryLetterTopic.set(oldRetryLetterTopic);
+                        }
+                    })
+                    .thenCompose(__ -> client.getPartitionedTopicMetadata(oldDeadLetterTopic))
```

Comment:
It's better to make `client.getPartitionedTopicMetadata(oldRetryLetterTopic)` and `client.getPartitionedTopicMetadata(oldDeadLetterTopic)` concurrent.

Reply:
yes, right. fixed.